mava/components/tf/losses/sequence.py

```diff
@@ -27,45 +28,71 @@ def recurrent_n_step_critic_loss(
 
+    # d: discount * done
+    # bootstrap_n=1 is the normal return of Q_t-1 = R_t-1 + d * Q_t
+    # bootstrap_n=seq_len is the Q_t_1 = discounted sum of rewards return
+
     seq_len = len(rewards[0])
-    assert bootstrap_n < seq_len
-    if type(q_values) != DiscreteValuedDistribution:
-        check_rank([q_values, target_q_values, rewards, discounts], [2, 2, 2, 2])
+    assert 0 < bootstrap_n <= seq_len
 
-    # Construct arguments to compute bootstrap target.
-    if type(q_values) != DiscreteValuedDistribution:
-        q_tm1 = q_values[:, 0:-bootstrap_n]
-        q_t = target_q_values[:, bootstrap_n:]
+    if not isinstance(q_values, DiscreteValuedDistribution):
+        # TODO (dries): Implement this test for MAD4PG as well.
+        check_rank([q_values, target_q_values, rewards], [2, 2, 2])
 
-        q_tm1, _ = combine_dim(q_tm1)
-        q_t, _ = combine_dim(q_t)
+    # The last values that rolled over do not matter because a
+    # mask is applied to it.
+    if not isinstance(q_values, DiscreteValuedDistribution):
+        # Construct arguments to compute bootstrap target.
+        q_tm1, _ = combine_dim(q_values)
+        q_t, _ = combine_dim(tf.roll(target_q_values, shift=-bootstrap_n, axis=1))
     else:
         q_tm1 = q_values
-        q_tm1.cut_dimension(axis=1, end=-bootstrap_n)
+        q_t = copy.copy(q_values)
 
-        q_t = target_q_values
-        q_t.cut_dimension(axis=1, start=bootstrap_n)
-    n_step_rewards = rewards[:, :-bootstrap_n]
-    n_step_discount = discounts[:, :-bootstrap_n]
-    for i in range(1, bootstrap_n + 1):
-        n_step_rewards += rewards[:, i : i + seq_len - bootstrap_n]
+        # Roll the logits inside the tfp distribution
+        # Question (dries): Is there a more elegant way to do this that working with
+        # _logits directly?
+        last_dim_shape = (q_t._logits.shape[-1],)
+        reshaped_logits = tf.reshape(q_t._logits, (rewards.shape + last_dim_shape))
+        rolled_logits = tf.roll(reshaped_logits, shift=-bootstrap_n, axis=1)
+        q_t._logits, _ = combine_dim(rolled_logits)
 
+    # Pad the rewards so that rewards at the end can also be calculated.
+    r_shape = rewards.shape
+    zeros_mask = tf.zeros(shape=r_shape[:-1] + (r_shape[-1] - bootstrap_n - 1,))
+    padded_rewards = tf.concat([rewards, zeros_mask], axis=1)
+    n_step_rewards = rewards
+    for i in range(1, bootstrap_n):
+        n_step_rewards += padded_rewards[:, i : i + seq_len] * math.pow(discount, i)
     n_step_rewards, _ = combine_dim(n_step_rewards)
-    n_step_discount, _ = combine_dim(n_step_discount)
 
-    critic_loss = loss_fn(
-        q_tm1,
-        n_step_rewards,
-        n_step_discount,
-        q_t,
-    )
+    # Create the end of episode mask.
+    ones_mask = tf.ones(shape=r_shape[:-1] + (r_shape[-1] - bootstrap_n,))
+    zeros_mask = tf.zeros(shape=r_shape[:-1] + (bootstrap_n,))
+    end_of_episode_mask, _ = combine_dim(tf.concat([ones_mask, zeros_mask], axis=1))
```

Comment:
What is purpose of this var? Seems like it always set to the same things (as opposed to reading it from replay). If this a flag to indicate end of episode, shouldn't this be read from replay?

Reply:
@KaleabTessera so we have two masks in the loss function. One is to mask zero-padded data and the other one (`end_of_episode_mask`) masks out the value function for where the data is not available for the target_value to bootstrap from. I am happy to have a call with you to further explain this. It would be nice if I can have a second pair of eyes on my code :slightly_smiling_face: